bouncer/src/repo/manipulator/modeloptimizer/repo_optimizer_multipart.cpp

```diff
@@ -399,75 +917,234 @@ bool MultipartOptimizer::isTransparent(
+	const repo::core::model::RepoScene* scene,
+	const BakedMeshNodes& bakedMeshNodes,
+	const std::set<repo::lib::RepoUUID>& groupMeshIds,
+	const repo::lib::RepoUUID& rootID,
+	repo::core::model::RepoNodeSet& mergedMeshes,
+	MaterialNodes& mergedMeshesMaterials,
+	MaterialUUIDMap& materialMap,
 	const bool isGrouped
 )
 {
-	if (!meshes.size()) return true;
+	// This method turns the submesh UUIDs into a set of supermesh MeshNodes 
+	// and duplicate material nodes. 
+	// They are added to the mergedMeshes and mergedMeshesMaterials arrays, to 
+	// be added to the stash graph when this method has been called for all 
+	// supermesh groups.
 
-	bool success = false;
-	auto sMesh = createSuperMesh(scene, meshes, matIDs, isGrouped);
-	if (success = sMesh)
+	if (!groupMeshIds.size())
 	{
-		auto sMeshWithParent = sMesh->cloneAndAddParent({ rootID });
-		sMesh->swap(sMeshWithParent);
-		mergedMeshes.insert(sMesh);
+		return true;
+	}
+
+	// First get all the meshes to build into the supermesh set. This snippet 
+	// returns the meshes baked into world space (where they should be when 
+	// combined).
 
-		repo::lib::RepoUUID sMeshSharedID = sMesh->getSharedID();
-		std::set<repo::lib::RepoUUID> currentMats;
-		for (const auto &map : sMesh->getMeshMapping())
+	std::vector<repo::core::model::MeshNode> nodes;
+	for (auto id : groupMeshIds)
+	{
+		auto range = bakedMeshNodes.equal_range(id);
+		for (auto pair = range.first; pair != range.second; pair++)
 		{
-			currentMats.insert(map.material_id);
+			nodes.push_back(pair->second);
 		}
+	}
 
-		/**
-		* FIXME: since there's multiple sMeshes, we need to make sure we only process the ones
-		* that are within the current sMesh. This is really inefficient to loop through all the matIDs
-		* but it will do for now. To do it properly we need a bi-directional map to find the original matID base
-		* on the new one
-		* This should all go away once we have proper texture support
-		*/
-		for (const auto matID : matIDs)
-		{
-			//Skip the material if it's not used in the current supermesh
-			if (currentMats.find(matID.second) == currentMats.end()) continue;
+	// Next partition them into sets that should form the supermeshes
+
+	std::vector<std::vector<repo::core::model::MeshNode>> clusters;
+	clusterMeshNodes(nodes, clusters);
+
+	// Build the actual supermesh nodes that hold the combined or split geometry from the sets
+
+	std::vector<repo::core::model::MeshNode*> supermeshes;
+	for (auto cluster : clusters)
+	{
+		createSuperMeshes(scene, cluster, materialMap, isGrouped, supermeshes);
+	}
+
+	// Create inverse lookup that goes from a re-mapped stash Id to the 
+	// original Id so processSupermeshMaterials can find the node to
+	// copy (when necessary).
+
+	MaterialUUIDMap stashIdToOriginalId;
+	for (const auto mapping : materialMap)
+	{
+		stashIdToOriginalId[mapping.second] = mapping.first;
```

Comment:
I don't think you need this reverse mapping?

Looking at `processSupermeshMaterials` you could do the same with just looping the original mapping - this feels a bit redundant

Reply:
With the refactor of processSupermeshMaterials it is especially redundant! The loop goes over the UUID map now which is much more efficient.